src/expr/cardinality_constraint.cpp

```diff
@@ -0,0 +1,108 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * The cardinality constraint operator
+ */
+
+#include "expr/cardinality_constraint.h"
+
+#include <iostream>
+
+#include "expr/type_node.h"
+
+namespace cvc5 {
+
+CardinalityConstraint::CardinalityConstraint(const TypeNode& type,
+                                             const Integer& ub)
+    : d_type(new TypeNode(type)), d_ubound(ub)
+{
+  AlwaysAssert(type.isSort())
+      << "Unexpected cardinality constraints for " << type;
+}
+
+CardinalityConstraint::CardinalityConstraint(const CardinalityConstraint& other)
+    : d_type(new TypeNode(other.getType())), d_ubound(other.getUpperBound())
+{
+}
+
+CardinalityConstraint::~CardinalityConstraint() {}
+
+const TypeNode& CardinalityConstraint::getType() const { return *d_type; }
+
+const Integer& CardinalityConstraint::getUpperBound() const { return d_ubound; }
+
+bool CardinalityConstraint::operator==(const CardinalityConstraint& cc) const
+{
+  return getType() == cc.getType() && getUpperBound() == cc.getUpperBound();
+}
+
+bool CardinalityConstraint::operator!=(const CardinalityConstraint& cc) const
+{
+  return !(*this == cc);
+}
+
+std::ostream& operator<<(std::ostream& out, const CardinalityConstraint& cc)
+{
+  return out << "fmf.card(" << cc.getType() << ", " << cc.getUpperBound()
+             << ')';
+}
+
+size_t CardinalityConstraintHashFunction::operator()(
+    const CardinalityConstraint& cc) const
+{
+  return std::hash<TypeNode>()(cc.getType())
+         * IntegerHashFunction()(cc.getUpperBound());
```

Comment:
It'd be slightly better to use our [`PairHashFunction`](https://github.com/cvc5/cvc5/blob/master/src/util/hash.h#L60-L66) util to combine the hashes.

Reply:
What is the code I'd use?

Reply:
It should be possible to do something like:

```
using CardinalityConstraintHashFunction = PairHashFunction<TypeNode, Integer, std::hash<TypeNode>, IntegerHashFunction>;
```

instead of declaring and implementing the class.